dali/kernels/dynamic_scratchpad_test.cc

```diff
@@ -0,0 +1,162 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/kernels/dynamic_scratchpad.h"  // NOLINT
+#include <gtest/gtest.h>
+#include <algorithm>
+#include <iostream>
+#include <chrono>
+#include <random>
+#include <vector>
+#include "dali/core/cuda_utils.h"
+#include "dali/core/cuda_stream_pool.h"
+#include "dali/core/mm/memory.h"
+
+namespace dali {
+namespace kernels {
+namespace test {
+
+/**
+ * @brief Tests basic dynamic scratchpad functioning
+ *
+ * This test checks that:
+ * - the memory is usable and accessible on the right backend
+ * - the pinned memory block is released in stream order, which prevents
+ *   immediate reuse if the stream is still running
+ * - it makes multiple attempts to catch the stream still running
+ */
+TEST(DynamicScratchpad, BasicTest) {
+  const int N = 64 << 10;  // 64 KiB
+
+  std::vector<char> in(N);
+  for (int i = 0; i < N; i++)
+    in[i] = i + 42;  // so it doesn't start or end with 0
+
+  auto stream = CUDAStreamPool::instance().Get();
+  auto dev = mm::alloc_raw_unique<char, mm::memory_kind::device>(N);
+  int max_attempts = 1000;
+  bool was_running = false;
+  for (int attempt = 0; attempt < max_attempts; attempt++) {
+    char *pinned;
+    {
+      DynamicScratchpad scratch({}, AccessOrder(stream));
+      pinned = scratch.Allocate<mm::memory_kind::pinned, char>(N);
+      memcpy(pinned, in.data(), N);
+      CUDA_CALL(cudaMemcpyAsync(dev.get(), pinned, N, cudaMemcpyHostToDevice, stream));
+    }
+    auto out = mm::alloc_raw_unique<char, mm::memory_kind::pinned>(N);
+    bool running = false;
+    if (was_running) {
+      CUDA_CALL(cudaStreamSynchronize(stream));
+    } else {
+      running = cudaStreamQuery(stream) == cudaErrorNotReady;
+      if (running)
+        was_running = true;
+    }
+    ASSERT_TRUE(out.get() + N < pinned || out.get() >= pinned + N || !running);
+    CUDA_CALL(cudaMemcpyAsync(out.get(), dev.get(), N, cudaMemcpyDeviceToHost, stream));
+    CUDA_CALL(cudaStreamSynchronize(stream));
+    ASSERT_EQ(memcmp(in.data(), out.get(), N), 0);
+    if (was_running && !running)
+      break;
+  }
+  if (!was_running)
+    std::cerr << "Warning: Test incomplete - the stream was never caught still running"
+              << std::endl;
+}
+
+TEST(DynamicScratchpad, Perf) {
+  std::poisson_distribution size_dist(1024);  // 1 KiB average
+  int max_size = 64 << 20;  // 64 MiB max
+  std::uniform_int_distribution<> num_dist(1, 100);
+
+  std::mt19937_64 rng(1234);
+
+  auto stream1 = CUDAStreamPool::instance().Get();
+  auto stream2 = CUDAStreamPool::instance().Get();
+  cudaStream_t streams[] = { stream1, stream2 };
+
+  int max_attempts = 100000;
+
+  const int nkinds = static_cast<int>(mm::memory_kind_id::count);
+  std::vector<double> alloc_times[nkinds];
+  std::vector<double> destroy_times;
+  for (auto &v : alloc_times)
+    v.reserve(max_attempts*10);
```

Comment:
```suggestion
    v.reserve(max_attempts*1024);
```
I think we do on average 1024 allocations per attempt.

Reply:
Yes, 100. I confused it with size_dist. So average would be like 50 and 100 max.